js/src/data/reducer.test.js

```diff
@@ -0,0 +1,641 @@
+/**
+ * External dependencies
+ */
+import { cloneDeep, set, get, isPlainObject } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import reducer from './reducer';
+import TYPES from './action-types';
+
+// Copied from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze
```

Comment:
Amazing work adding the tests 🥳 

❓ 
I understand that we use that deepFreeze + assertRefs function in order to test the un-mutability of the reducer. 

Checking here I don't see it as standard in the Redux testing docs. Do we usually do this in WP/WC testing?
https://redux.js.org/usage/writing-tests

Not judging but it increases the complexity of the tests. I would definitely at least hide deepFreeze and attachRef in some kind of test utils separated file. 

- We can re-use if needed in other reducers 
- Hides the complexity for the next developer who want to work on this tests.

Reply:
> Checking here I don't see it as standard in the Redux testing docs. Do we usually do this in WP/WC testing?

I viewed some WC repos and they usually only do the `deepFreeze` checks, but not reference checks. I think it's probably not very common to do this.

When dealing with #270, I wanted a way to ensure that the reference change was indeed corrected by the new implementation. So I chose to write a test to verify it. If the issue didn't exist in the first place, I don't think it would be necessary to do this special reference check, but I think it's an okay consideration to create a test to ensure that it doesn't occur again (at least for a while) based on the fact that it has already occurred.

Since this is a react-redux-specific convention, whether we should default to developers knowing the convention, or rely on test-assisted validation, I think this is a separate 📜  discussion.

As for the tests that have been written, I realized it's a bit complex, and it might take a bit of time to read the code comments and even the implementation code to understand how to use them when adding new tests, but I tend to keep them and observe them for a while before deciding whether to remove them.

> Not judging but it increases the complexity of the tests. I would definitely at least hide deepFreeze and attachRef in some kind of test utils separated file.
> 
> * Hides the complexity for the next developer who want to work on this tests.

I tried to isolate as much of its complexity as possible behind `state.assertConsistentRef()` and its code comments, but if developers want to understand what `assertConsistentRef`, `deepFreeze` or `attachRef` actually do, then I don't think the complexity would be hidden whether they are separated files or not.

> * We can re-use if needed in other reducers

Since there's currently only one reducer, I tend to move them out when we need to reuse them.

Reply:
> When dealing with #270, I wanted a way to ensure... 

Gotcha, since the issue is pretty old and with tons of discussion and I don't want to delay this PR by reading and understanding all the history, plus the fact that the test are 600 lines long. I would agree to go ahead with it and then discuss further in  #270 or a new issue if necessary.

> As for the tests that have been written, I realized it's a bit complex, and it might take a bit of time to read the code ... 

Same as before, let's go ahead for now and discuss if needed in next Cooldown if there is a better way to handle this.

> I tried to isolate as much of its complexity as possible behind state.assertConsistentRef() and its code comments, but if developers want to understand what assertConsistentRef, deepFreeze or attachRef actually do, then I don't think the complexity would be hidden whether they are separated files or not.

This I still disagree, since there is a complex solution of around 100 lines.

"hiding" it behind an aux/helper file would help in terms of readability and focus on the important part (the tests) that are very neat built btw. 

What I expected was to have an import at the beginning with the next abstractions in utils:

- deepFreeze
- assertConsistentRef
- prepareState 

Besides that, It would be good to have the defaultState imported from DEFAULT_STATE constant instead of hardcoding it in the beforeEach statement. 

Then we can just check the expressive tests in the describe sections as you did :) In the end I want just to trust in the function implementations like `assertConsistentRef` and check it by clicking on the import only if I want to go deeper. 

And again, even tho there is no other reducer it's always nice to be prepared for so if needed. But in this case it's more about readability and fragmentation than reusability. 

thoughts?¿?¿?

Reply:
Thanks for the explanation. I guess now I understand you're focusing on readability, I thought you meant that moving those functions to a separate file would hide their complexity.

> This I still disagree, since there is a complex solution of around 100 lines.
> "hiding" it behind an aux/helper file would help in terms of readability and focus on the important part (the tests) that are very neat built btw.

Readability could sometimes be affected by subjectivity, preference, or editors' usage habits. For example, I'm used to reading and grasping the details of code, so it's sometimes better for me to have highly interdependent and non-shared functions in the same file. So I don't need to open many files and switch back and forth between them. If I don't want to be distracted by pieces of code for a while, I will use editor to fold them up.

Of course, when there are too many lines in a single file, readability is getting worse, but how many lines are too many is a matter of opinion. I'm happy to move them because it's a ❓ level concern for you. I opened a separate PR #1153 to move it, not only for the reasons mentioned in that PR description but also to save the extra efforts of dealing with the conflict between #1151.

Reply:
Agree that how much is too much is subjective... however the simpler the better IMO. 

As suggested let's discuss elaborate in PR #1153